ballet/update.py

```diff
@@ -0,0 +1,84 @@
+import funcy
+import git
+import json
+import os
+import tempfile
+import yaml
+
+from cookiecutter.main import cookiecutter
+from cookiecutter.prompt import prompt_for_config
+
+from ballet.compat import pathlib, safepath
+from ballet.util.log import logger
+from ballet.quickstart import generate_project
+
+PROJECT_CONTEXT_PATH = (
+    pathlib.Path(__file__).resolve().parent.joinpath('project_template', 'cookiecutter.json')
+)
+
+REPLAY_PATH = (
+    pathlib.Path.home().joinpath('.cookiecutter_replay', 'project_template.json')
+)
+
+def _create_replay(tempdir, context):
+    tempdir = pathlib.Path(tempdir)
+    slug = context['cookiecutter']['project_slug']
+    old_context = None
+    try:
+        # replace the current replay with our own
+        with open(REPLAY_PATH, 'r') as replay_file:
+            old_context = json.load(replay_file)
+        with open(REPLAY_PATH, 'w') as replay_file:
+            json.dump(context, replay_file)
+        generate_project(replay=True, output_dir=tempdir)
+    except:
+        # we're missing keys, figure out which and prompt
+        with open(PROJECT_CONTEXT_PATH) as context_json:
+            new_context = json.load(context_json)
+        new_keys = set(new_context.keys()) - set(context['cookiecutter'].keys())
+        new_context_config = {'cookiecutter': funcy.project(new_context, new_keys)}
+        new_context = prompt_for_config(new_context_config)
+        context['cookiecutter'].update(new_context)
+
+        with open(REPLAY_PATH, 'w') as replay_file:
+            json.dump(context, replay_file)
+        generate_project(replay=True, output_dir=tempdir)
```

Comment:
Failure mode: This might fail as well, it would not be caught, the finally block would be called, and the old context would not be restored.

How about something along the lines of `def _create_replay(tempdir, context, old_context)` which can then be re-called into from this except block. and separate an `except KeyError` for cases where we are missing keys and an `except Exception` which catches all other "un-recoverable" errors and makes sure to restore the original cookiecutter_replay file.

Reply:
if this is not that clear we can discuss further

Reply:
I think trying to do two replays was weird as a way to check for errors, so I opted to check for new context pre-emptively and so there's no critical logic inside the except block now.